### PR TITLE
Harden web server security

### PR DESCRIPTION
**Pull request by deiteris**
* Breaking change: set host to `127.0.0.1` by default to prevent external requests for local-only instances.
* Breaking change: allow only local origins when allowed origins are not specified in CLI.
* Breaking change: add Origin header check to prevent requests execution from foreign origins.
* Add `--host` argument to specify host interface.
* Add `--allowed-origins` argument to specify a list of origins from which access is allowed. For example, `--allowed-origins https://www.google.com --allowed-origins https://google.com`. Access is allowed from 127.0.0.1 and localhost origins by default.
* Update print messages with URL. `127.0.0.1` is not allowed to have access to AudioContext in Chrome.
* Remove permissive CORS middleware.

Addresses https://github.com/w-okada/voice-changer/issues/1114.

### CODE REVIEW

**Comment by deiteris** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Venryx** [comment]
I've been doing some testing on your [harden-security branch](https://github.com/deiteris/voice-changer/tree/harden-security), and it seems to work well!

Some of the things I tried from another origin (website, and another localhost port), which were successfully blocked: (tested with server in both https and non-https mode)
* `fetch(".../update_settings", {method: "POST", body: "..."})`
* `fetch(".../update_settings", {method: "POST", mode: "no-cors", body: "..."})`
* `fetch(".../update_settings", {method: "POST", referrerPolicy: 'no-referrer', body: "..."})`
* `fetch(".../update_settings", {method: "POST", mode: "no-cors", referrerPolicy: 'no-referrer', body: "..."})`
* `fetch(".../info", {method: "GET"})`
* `fetch(".../info", {method: "GET", mode: "no-cors"})`
* `fetch(".../info", {method: "GET", referrerPolicy: 'no-referrer'})`
* `fetch(".../info", {method: "GET", mode: "no-cors", referrerPolicy: 'no-referrer'})`

The only thing I was able to do, is perform "GET" requests to the server (eg. `/info` and `/onnx`, with my observing a log-message of "No pyTorch filepath" on the server each time the `/onnx` endpoint was fetched), when:
* Using the Chrome browser. (Firefox blocks `fetch` from doing even these "GET" requests, albeit you can workaround using iframes, etc.)
* The request body is empty. (browser blocks providing a body in GET requests)
* The response's contents are unable to be read. (browser makes the response "opaque" by replacing it with an empty string)

So, the above is the standard/expected behavior. That said, the one minor "negative" of the "GET" requests being allowed, is that they do technically allow websites to observe whether the W-Okada Voice Changer app is running (on Chrome at least):
```
// if voice-changer is running, this returns an empty string immediately; else, it ends up timing out instead
await (await fetch("http://localhost:18888/info", {
    method: "GET", mode: "no-cors",
})).text()
```

That's not that big a deal though; only real drawback I can see is some embarrassment if someone using a voice-changer has that revealed by opening a link that detects the app being active. While not 100% ideal, I don't see any easy way to solve this last part. (Since browsers, eg. Chrome, just aren't designed to be able to _outright block_ opaque, cross-origin `GET` requests; while theoretically blocking these GET requests could happen in the server, eg. by blocking if the `referrer` header is from another domain, it might be difficult to exactly mimic the type of response that is consistent with the app not running at all, and I don't want to hold up a major security improvement over a minor detail like this.)

In summary: This pull-request does appear to offer a major improvement in security from how things used to be. Thank you for your work fixing this, and making the list of allowed-origins configurable (as this retains flexibility for some custom setups). I hope this can be merged soon!

**Comment by deiteris** [comment]
> The only thing I was able to do, is perform "GET" requests to the server (eg. `/info` and `/onnx`, with my observing a log-message of "No pyTorch filepath" on the server each time the `/onnx` endpoint was fetched), when:
> 
> * The server is launched in non-https mode. (https-mode blocks cross-origin GET requests with the [seemingly non-js-visible] error `net::ERR_CERT_AUTHORITY_INVALID`)
> * The request body is empty. (browser blocks providing a body in GET requests)
> * The response's contents are unable to be read. (browser makes the response "opaque" by replacing it with an empty string)

Were you able to actually read those requests? With the latest commit on that branch, I'm getting error 400 from other origins no matter which parameters I pass. This is what is expected:
image
image

**Comment by Venryx** [comment]
> Were you able to actually read those requests? With the latest commit on that branch, I'm getting error 400 from unknown origins no matter which parameters I pass.
> 
> So even reading should not be possible in case you're trying to access from unknown origin.

Correct, I was not able to actually read those requests. (I simply used the "empty string" opaque response to observe that the voice-changer app was active, since if it weren't, the request would just error/timeout instead; which is expected, not so serious, and hard to 100% solve anyway)

**Comment by Venryx** [comment]
@w-okada 重要: このプル リクエストに関するフィードバックを提供していただけますか?マージを検討する前に解決してほしい問題はありますか?

現状では、このプログラムをバックグラウンドで実行したままにしておくのは信用できません。マシン上で任意の Python コードを実行しているランダムな Web サイトにアクセスしたままになってしまうからです。 (具体的には、w-okada プログラム/バックエンドへの呼び出しを通じて、pickle エクスプロイトによってアクティブ化された悪意のあるコードを含むカスタムの「音声モデル」をアップロードしてアクティブ化することによって、その API は現在どの Web サイト/ドメインからでもアクセスでき、完全に完了します)沈黙）

[the above is the Japanese translation, in case w-okada is time-constrained and only sees the email notification, which may not offer automatic translation to make a read-through convenient; english version below]

Can you please provide feedback on this pull request? Is there something wrong with it that you'd like addressed before considering merging it?

As it stands, I have trouble being trusting of leaving this program running in the background, as it leaves me open to random websites running arbitrary code on my machine. (specifically: by uploading and activating a custom "voice model" containing malicious code, activated by means of the [pickle exploit](https://arstechnica.com/security/2024/03/hugging-face-the-github-of-ai-hosted-code-that-backdoored-user-devices/), through calls to the w-okada program/backend, whose API can currently [be accessed](https://github.com/w-okada/voice-changer/issues/1114) from any website/domain and in complete silence)

**Comment by w-okada** [comment]
Thanks for your PR. 
Your proposal is likely to be very important. However, we cannot immediately determine how significant the changes will impact. We will proceed with the merge, but please note there is a possibility that we will roll it back depending on the effects.
